lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -1880,54 +2050,38 @@ bool EarlyFunctionAnalysis::runOnModule(Module &M) {
                                  nullptr);
   FunctionAnalysisResults Properties(std::move(DefaultSummary));
 
-  // Instantiate a CFEPAnalyzer object
-  using CFEPA = CFEPAnalyzer<FunctionAnalysisResults>;
-  CFEPA Analyzer(M, &GCBI, Properties, ABIRegisters);
+  // Cache pre-population
+  Properties.importModel(M, ABIRegisters, Binary);
 
-  // Interprocedural analysis over the collected functions in post-order
-  // traversal (leafs first).
-  while (!CFEPQueue.empty()) {
-    BasicBlockNode *EntryNode = CFEPQueue.pop();
-    revng_log(EarlyFunctionAnalysisLog,
-              "Analyzing Entry: " << EntryNode->BB->getName());
+  // Instantiate a FunctionEntrypointAnalyzer object
+  using FEA = FunctionEntrypointAnalyzer<FunctionAnalysisResults>;
+  FEA Analyzer(M, &GCBI, ABIRegisters, &EntrypointsQueue, Properties, Binary);
 
-    // Intraprocedural analysis
-    FunctionSummary AnalysisResult = Analyzer.analyze(EntryNode->BB);
-    bool Changed = Properties.registerFunction(getBasicBlockPC(EntryNode->BB),
-                                               std::move(AnalysisResult));
+  // Either recover the control-flow graph of the program and serialize it to
+  // LLVM IR, or schedule a full ABI analysis for each function detected
+  // (control-flow graph recovery included).
+  if (!ShouldAnalyzeABI) {
+    Analyzer.recoverCFG();
+  } else {
+    // Interprocedural analysis over the collected functions in post-order
+    // traversal (leafs first).
+    Analyzer.runInterproceduralAnalysis();
 
-    // If we got improved results for a function, we need to recompute its
-    // callers, and if a caller turns out to be fake, the callers of the fake
-    // function too.
-    if (Changed) {
-      UniquedQueue<BasicBlockNode *> FakeFunctionWorklist;
-      FakeFunctionWorklist.insert(EntryNode);
+    // Propagate results between call-sites and functions
+    Analyzer.interproceduralPropagation();
 
-      while (!FakeFunctionWorklist.empty()) {
-        BasicBlockNode *Node = FakeFunctionWorklist.pop();
-        for (auto *Caller : Node->predecessors()) {
-          if (Caller == RootNode)
-            break;
-
-          if (!Properties.isFakeFunction(getBasicBlockPC(Caller->BB)))
-            CFEPQueue.insert(Caller);
-          else
-            FakeFunctionWorklist.insert(Caller);
-        }
-      }
-    }
+    // Commit the results onto the model
+    Analyzer.finalizeModel();
   }
 
+  // Serialize function metadata, including the CFG, to the entry basic blocks
+  // in `root`.
+  serializeFunctionMetadata(M.getContext(), GCBI, Properties, Binary);
```

Comment:
Consider making this a method of `Analyzer`.

Reply:
Took the freedom not to move this method this time, I felt like the serialization of metadata is completely independent from the whole analysis of the function. Lemme know if you have strong opinions on this though, and I'll move it.